### PR TITLE
Use ReadOnlySpan and ReadOnlyMemory for reducing memory copy

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

**Comment by erikzhang** [comment]
With these changes, we should have very high performance when executing large memory copy instructions, such as `PUSH*`, `SUBSTR`, `LEFT`, `RIGHT`, and so on.

File: src/neo-vm/Types/Integer.cs [hunk 1]
```diff
@@ -39,17 +39,24 @@ public override bool GetBoolean()
...
+        {
+            return value.GetByteCount();
+        }
+#else
         private int _length = -1;
         public override int GetByteLength()
         {
             if (_length == -1)
-                _length = value.ToByteArray().Length;
+                _length = value.GetByteCount();
```
Comment (shargon) [inline src/neo-vm/Types/Integer.cs]: whi with `NETCOREAPP` is not cached?

Comment (erikzhang) [reply src/neo-vm/Types/Integer.cs]: In netcoreapp2.1+, `GetByteCount()` is a member method of `BigInteger`, and it returns the length directly from the internal field.

**Comment by erikzhang** [comment]
@shargon Do you need more time to review it?

**Comment by shargon** [comment]
In general all of these changes improves the VM :)

```
BenchmarkDotNet=v0.11.4, OS=Windows 10.0.17134.648 (1803/April2018Update/Redstone4)
Intel Core i7-6700HQ CPU 2.60GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
Frequency=2531253 Hz, Resolution=395.0613 ns, Timer=TSC
.NET Core SDK=2.2.104
  [Host]     : .NET Core 2.1.9 (CoreCLR 4.6.27414.06, CoreFX 4.6.27415.01), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.9 (CoreCLR 4.6.27414.06, CoreFX 4.6.27415.01), 64bit RyuJIT
```

| Method |   OpCodes |     Mean |    Error |   StdDev |   Median |      Min |      Max | Rank |
|------- |---------- |---------:|---------:|---------:|---------:|---------:|---------:|-----:|
| PR-104 | VERIFY*1K | 659.9 ms | 13.13 ms | 17.52 ms | 649.5 ms | 638.1 ms | 689.3 ms |    1 |
| 2.4.0  | VERIFY*1K | 755.0 ms | 14.99 ms | 35.62 ms | 669.8 ms | 834.9 ms | 756.7 ms |    2 |

| Method | OpCodes |     Mean |     Error |    StdDev |      Min |      Max |   Median | Rank |
|------- |-------- |---------:|----------:|----------:|---------:|---------:|---------:|-----:|
| PR-104 | SHA1*1K | 2.109 ms | 0.0655 ms | 0.0673 ms | 2.054 ms | 2.242 ms | 2.079 ms |    1 |
| 2.4.0  | SHA1*1K | 3.801 ms | 0.0453 ms | 0.0423 ms | 3.729 ms | 3.880 ms | 3.805 ms |    2 |

| Method | OpCodes |     Mean |    Error |   StdDev |      Min |      Max |   Median | Rank |
|------- |-------- |---------:|---------:|---------:|---------:|---------:|---------:|-----:|
| PR-104 |  NOP*1K | 282.3 us | 3.316 us | 3.102 us | 276.6 us | 287.4 us | 282.6 us |    2 |
| 2.4.0  |  NOP*1K | 249.7 us | 14.35 us | 42.30 us | 273.1 us | 149.0 us | 290.7 us |    1 |

| Method |   OpCodes |     Mean |     Error |    StdDev |      Min |      Max |   Median | Rank |
|------- |---------- |---------:|----------:|----------:|---------:|---------:|---------:|-----:|
| PR-104 | FACTORIAL | 11.30 us | 0.3018 us | 0.8899 us | 9.726 us | 12.67 us | 11.15 us |    1 |
| 2.4.0  | FACTORIAL | 12.07 us | 0.1568 us | 0.1467 us | 11.88 us | 12.40 us | 12.09 us |    2 |

| Method |   OpCodes |     Mean |     Error |   StdDev  |   Median |      Min |      Max | Rank |
|------- |---------- |---------:|----------:|----------:|---------:|---------:|---------:|-----:|
| PR-104 | FIBONACCI | 10.29 us | 0.5893 us | 1.7380 us | 10.96 us | 6.390 us | 12.11 us |    1 |
| 2.4.0  | FIBONACCI | 10.57 us | 0.2440 us | 0.7193 us | 8.999 us | 11.81 us | 10.52 us |    2 |

| Method |         OpCodes |     Mean |    Error |   StdDev |      Min |        Max |   Median | Rank |
|------- |---------------- |---------:|---------:|---------:|---------:|-----------:|---------:|-----:|
| PR-104 | (PUSH0+DROP)*1K | 839.5 us | 36.26 us | 106.9 us | 618.6 us | 1,025.6 us | 857.8 us |    1 |
| 2.4.0  | (PUSH0+DROP)*1K | 894.0 us | 18.61 us | 54.86 us | 763.3 us |  971.70 us | 904.6 us |    2 |

I hope to approve it tomorrow, I spend more time than expected with the benchmarks.

**Pull request merged**